vm/src/macros.rs

```diff
@@ -38,30 +38,35 @@ macro_rules! py_class {
     };
     ( $ctx:expr, $class_name:expr, $class_base:expr, $flags:expr, { $($name:tt => $value:expr),* $(,)* }) => {
         {
-            let py_class = $ctx.new_class($class_name, $class_base, $crate::slots::PyTpFlags::DEFAULT | $flags);
-            $crate::extend_class!($ctx, &py_class, { $($name => $value),* });
+            #[allow(unused_mut)]
+            let mut slots: $crate::slots::PyClassSlots = ($crate::slots::PyTpFlags::DEFAULT | $flags).into();
+            $($crate::py_class!(@extract_slots($ctx, &mut slots, $name, $value));)*
+            let py_class = $ctx.new_class($class_name, $class_base.clone(), slots);
+            $($crate::py_class!(@extract_attrs($ctx, &py_class, $name, $value));)*
+            $ctx.add_tp_new_wrapper(&py_class);
             py_class
         }
     };
+    (@extract_slots($ctx:expr, $slots:expr, (slot $slot_name:ident), $value:expr)) => {
+        $slots.$slot_name = Some(
+            $crate::function::IntoPyNativeFunc::into_func($value)
+        );
+    };
+    (@extract_slots($ctx:expr, $class:expr, $name:expr, $value:expr)) => {};
+    (@extract_attrs($ctx:expr, $slots:expr, (slot $slot_name:ident), $value:expr)) => {};
+    (@extract_attrs($ctx:expr, $class:expr, $name:expr, $value:expr)) => {
+        $class.set_str_attr($name, $value);
+    };
```

Comment:
Sort of; I know that `tokio` does that with [their macros](https://docs.rs/tokio/0.2.22/tokio/macro.select.html); I think it keeps the logic encapsulated so that when there's a `@extract_slots` inside you know that it's in the same macro definition.